File: ocaml/basics/fibonacci.ml
(* Print the first 20 Fibonacci numbers, accumulated by a tail-recursive loop. *)
(* Run: ocaml fibonacci.ml *)

let fibonacci count =
  let rec loop i current next acc =
    if i = count then List.rev acc
    else loop (i + 1) next (current + next) (current :: acc)
  in
  loop 0 0 1 []

let () =
  fibonacci 20 |> List.map string_of_int |> String.concat " " |> print_endline

File: erlang/basics/fibonacci.erl
%% Print the first 20 Fibonacci numbers, accumulated by a tail-recursive loop.
%% Run: escript fibonacci.erl

-module(fibonacci).
-export([main/1]).

fibonacci(Count) ->
    fibonacci(Count, 0, 1, []).

fibonacci(0, _Current, _Next, Acc) ->
    lists:reverse(Acc);
fibonacci(Count, Current, Next, Acc) ->
    fibonacci(Count - 1, Next, Current + Next, [Current | Acc]).

main(_) ->
    Values = [integer_to_list(V) || V <- fibonacci(20)],
    io:format("~s~n", [string:join(Values, " ")]).
